1) При прерывании скрипта выводится сообщение "Аварийное завершение". В сценарии мы устанавливаем данную реакцию на сигнал прерывания SIGINT при помощи trap. Этот сигнал посылается нажатием Ctrl+C.
2) $$ возвращает идентификатор bash, а /proc/self ссылается на директорию процесса, который обратился к этой ссылке (в данном случае процесс команды ls, можно в этом удостовериться поставив в конце команды & и получив идентификатор только что запущенного процесса ls)
3) 0 - stdin, 1 - stdout, 2 - stderr
4) Стандартный вывод (1 - stdout) перенаправляется в файл /tmp/ls.out, а вывод ошибок (2 - stderr) в файл /tmp/ls.err. Эти изменения зафиксированы в выводе результата выполненной команды (/tmp/ls.out).
5) ls -l /proc/self/fd > /tmp/ls.out 2> /tmp/ls.err < /tmp/ls.in
Стандартный ввод (0 - stdin) перенаправлен в файл /tmp/ls.in, в этом можно убедиться, посмотрев файл /tmp/ls.out.
6) При выполнении exec ps -l от имени shell процесс ps заменяет процесс sh. Без exec у нас shell создал бы дочерний процесс ps, и оба процесса (sh и ps) были бы запущены, что и отображается при выводе на экран.
7) 3 дескриптор перенаправлен на файл test.out. Его содержимое:
Test3
Test333
pos указывает на позицию каретки при записи в файл, или на количество символов в выводе. Буквы и цифры составляют вместе 12 символов. Также прибавляются два символа переноса строки, автоматически добавляемые командой echo.
8) Если обращаться к файлу через имя - будет писать ошибку, через дескриптор - просто ничего не выводится. Наверное, файл считается удаленным, когда удалены все ссылки на него и ни один дескриптор процесса не использует его. И мы удалили не сам файл, а его имя/ссылку на область памяти...
